BitmapData draw & sync optimisations

Dinnerbone opened the pull request:
Main changes are:
1 - We only read back the area that (might have) changed. If you did a draw that only affected a 10x10 region of a 100x100 bitmap, only that 10x10 is read back
2 - We now allocate a pooled buffer for reading back from a BitmapData.draw. Number of allocations should be way down, and those are slow on webgl. The size of the buffer depends on the previous optimisation, not the whole texture
3 - Similar to 1 but for uploads, we only update the changed area of a texture when uploading it back to gpu
4 -  Moved all shared bitmapdata operations into their own file so we can make shared optimisations surrounding the actual BitmapData usage without duplicating in each VM
5 -  Fixed a number of places where we weren't syncing from cpu->gpu
6 - Fixed a number of places where we were syncing from gpu->cpu and didn't need to
7 - A number of methods now only do the gpu->cpu sync if they're going to read within the dirty area; otherwise they use the cached area
8 - Completely skip any BitmapData.draw() calls that are culled or otherwise won't affect any pixel



This should make any game that uses bitmapdata.draw much faster, especially on web.

SN902 commented:
`thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Queue::write_texture
    copy of X 0..641 would end up overrunning the bounds of the Destination texture of X size 640

', C:\Users\Sasha\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-0.15.1\src\backend\direct.rs:3024:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
AVM2 stack trace:
        at flash.display::BitmapData/flash::display::BitmapData::draw()
        at gxtkGraphics/DrawSurface2()
        at MethodInfo-8()
        at bb_guardic_return_2_GameObject/m_Draw()
        at bb_guardic_return_MyGame/m_OnRender()
        at bb_app_AppDevice/OnRender()
        at gxtkApp/InvokeOnRender()
        at MethodInfo-113()
        at MonkeyGame/onAddedToStage()`
        
[Guardicalogia_r4.zip](https://github.com/ruffle-rs/ruffle/files/11043898/Guardicalogia_r4.zip)

Dinnerbone commented:
@SN902 that's fixed now but it looks like that game suffers from the new pooling method, I think the game leaks bitmapdatas so we can never recycle them :(

SN902 commented:
![image](https://user-images.githubusercontent.com/38821724/227018748-c8eac16d-f153-41eb-89b7-684d85a02154.png)

`thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_buffer
    not enough memory left

', C:\Users\Sasha\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-0.15.1\src\backend\direct.rs:3024:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
AVM2 stack trace:
        at flash.display::BitmapData/flash::display::BitmapData::draw()
        at gxtkGraphics/DrawSurface2()
        at MethodInfo-8()
        at bb_guardic_return_2_GameObject/m_Draw()
        at bb_guardic_return_MyGame/m_OnRender()
        at bb_app_AppDevice/OnRender()
        at gxtkApp/InvokeOnRender()
        at gxtkApp/OnEnterFrame()`

SN902 commented:
> @SN902 that's fixed now but it looks like that game suffers from the new pooling method, I think the game leaks bitmapdatas so we can never recycle them :(

yeah :(

Dinnerbone commented:
Try now @SN902!

SN902 commented:
> Try now @SN902!

Guardicalogia fixed, awesome :)

Dinnerbone commented:
I added a bunch more improvements, namely:
- Moved all shared bitmapdata operations into their own file so we can make shared optimisations surrounding the actual BitmapData usage without duplicating in each VM
- Fixed a number of places where we weren't syncing from cpu->gpu
- Fixed a number of places where we were syncing from gpu->cpu and didn't need to
- A number of methods now only do the gpu->cpu sync if they're going to read within the dirty area; otherwise they use the cached area
- Completely skip any BitmapData.draw() calls that are culled or otherwise won't affect any pixel

SN902 commented:
new regression [Rainbow Dash Shooter Project.zip](https://github.com/ruffle-rs/ruffle/files/11068512/Rainbow.Dash.Shooter.Project.zip)

![image](https://user-images.githubusercontent.com/38821724/227709197-e081d5d0-6499-48fa-b4bf-391e2e51c5a5.png)

Dinnerbone commented:
Fixed the panic in Rainbow Dash Shooter Project. It's broken on master too but *mostly* doesn't crash and instead does wrong behaviour (usually skipping the draw). Now it shouldn't crash and should be correct.

Should also fix https://github.com/ruffle-rs/ruffle/issues/10264, https://github.com/ruffle-rs/ruffle/issues/10017, https://github.com/ruffle-rs/ruffle/issues/9923 and https://github.com/ruffle-rs/ruffle/issues/9875

Dinnerbone commented:
All the issues @SN902 found should be fixed, and we've got a few new tests for them. This PR should be final now.

@Herschel I'll look at merging the Rectangle classes in another PR, they'll likely need changes to things using them and this PR is big enough as it is.

Dinnerbone commented:
All the issues @SN902 found should be fixed, and we've got a few new tests for them. This PR should be final now.

@Herschel I'll look at merging the Rectangle classes in another PR, they'll likely need changes to things using them and this PR is big enough as it is.

Herschel commented:
Thank you! I won't have the bandwidth nor skill to give this a thorough review, but I can at least give it a skim this week. Otherwise I'll leave it up to you to merge when you think it's ready.

n0samu commented:
Fixes #10467, #9852, and #9970!

Herschel commented on core/src/bitmap/bitmap_data_operations.rs:
```diff
@@ -0,0 +1,1364 @@
+    y: u32,
+    color: Color,
+) {
+    if x >= target.width() || y >= target.height() {
```
Probably from prior to this PR, but are some of these missing a `disposed` check, or is the `disposed` check redundant? Seems like we should be consistent with whether we check or not.

Dinnerbone replied:
I'll move them out of here and put the onus on the caller, since AVM1 and AVM2 respond to disposed differently

Herschel commented on core/src/bitmap/bitmap_data.rs:
```diff
@@ -314,29 +307,54 @@ mod wrapper {
+                if area.intersects(read_area) {
+                    true
+                } else {
+                    tracing::info!("SKIPPED");
```
Debug spew

Dinnerbone replied:
Thanks, removed!

Herschel commented on render/src/bitmap.rs:
```diff
@@ -158,3 +159,145 @@ impl BitmapFormat {
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct PixelRegion {
+    pub min_x: u32,
+    pub min_y: u32,
+    pub max_x: u32,
+    pub max_y: u32,
```
Unimportant naming nit: Looks like `Rectangle` does `x_min` instead of `min_n`. Would be nice to use consistent naming convention here even if we end up merging them eventually.

Dinnerbone replied:
Renamed

Herschel reviewed (approved):
I gave it a quick once over posted some small nits but LGTM! Thank you!

Dinnerbone commented on render/wgpu/src/buffer_pool.rs:
```diff
@@ -100,67 +100,113 @@ struct GlobalsKey {
+        }
+
+        let (item, used_description) = if let Some((_, best)) = best {
+            guard.remove(best)
```
Good idea, order isn't important
